cli/js/compiler_api.ts

```diff
@@ -0,0 +1,395 @@
+  suppressExcessPropertyErrors?: boolean;
+
+  /** Suppress `noImplicitAny` errors for indexing objects lacking index
+   * signatures. */
+  suppressImplicitAnyIndexErrors?: boolean;
+
+  /** Specify ECMAScript target version. Defaults to `esnext`. */
+  target?:
+    | "es3"
+    | "es5"
+    | "es6"
+    | "es2015"
+    | "es2016"
+    | "es2017"
+    | "es2018"
+    | "es2019"
+    | "es2020"
+    | "esnext";
+
+  /** List of names of type definitions to include. Defaults to `undefined`. */
+  types?: string[];
+}
+
+/** Internal function to just validate that the specifier looks relative, that
+ * it starts with `./`. */
+function checkRelative(specifier: string): string {
+  return specifier.match(/^([\.\/\\]|https?:\/{2}|file:\/{2})/)
+    ? specifier
+    : `./${specifier}`;
+}
+
+/** The results of a transpile only command, where the `source` contains the
+ * emitted source, and `map` optionally contains the source map.
+ */
+export interface TranspileOnlyResult {
+  source: string;
+  map?: string;
+}
+
+/** Takes a set of TypeScript sources and resolves with a map where the key was
+ * the original file name provided in sources and the result contains the
+ * `source` and optionally the `map` from the transpile operation. This does no
+ * type checking and validation, it effectively "strips" the types from the
+ * file.
+ *
+ *      const results =  await Deno.transpileOnly({
+ *        "foo.ts": `const foo: string = "foo";`
+ *      });
+ *
+ * @param sources A map where the key is the filename and the value is the text
+ *                to transpile.  The filename is only used in the transpile and
+ *                not resolved, for example to fill in the source name in the
+ *                source map.
+ * @param options An option object of options to send to the compiler. This is
+ *                a subset of ts.CompilerOptions which can be supported by Deno.
+ *                Many of the options related to type checking and emitting
+ *                type declaration files will have no impact on the output.
+ */
+export function transpileOnly(
+  sources: Record<string, string>,
+  options?: CompilerOptions
+): Promise<Record<string, TranspileOnlyResult>> {
+  util.log("Deno.transpileOnly", { sources: Object.keys(sources), options });
+  const payload = {
+    sources,
+    options: options ? JSON.stringify(options) : undefined
+  };
+  return sendAsync(dispatch.OP_TRANSPILE, payload).then(result =>
+    JSON.parse(result)
+  );
+}
+
+/** Takes a root module name, any optionally a record set of sources. Resolves
+ * with a compiled set of modules.  If just a root name is provided, the modules
+ * will be resolved as if the root module had been passed on the command line.
+ *
+ * If sources are passed, all modules will be resolved out of this object, where
+ * the key is the module name and the value is the content.  The extension of
+ * the module name will be used to determine the media type of the module.
```

Comment:
So it effectively means that files compiled via CLI entry point can consume `content-type` header of downloaded module and fallback to file extension for media type, but compilation during runtime always uses file extension?

Reply:
Compilation where the sources are supplied.  If no sources are supplied, the module resolution, including supporting mime-types, is supported.